### DISCUSSION TITLE
emulation leak on new tab

### DISCUSSION THREAD

**Discussion by tnt2k** [kaliiiiiiiiii/Selenium-Profiles#50] [General]
Tell me how to inherit a profile when opening a new tab or when clicking on a link from a window with a profile?
For some reason, my profile flies when I open a new tab. Found a solution to this problem via 
```
driver.profiles.apply(profile)
``` 
and that results in an warnings
```
UserWarning: profile["options"] can't be applied when driver allready started
   warnings.warn('profile["options"] can\'t be applied when driver allready started')
```
, this warning is understandable since the proxy options are specified, but for some reason when trying to apply
```
driver.profiles.apply(profile["cdp" ])
``` 
getting the error
```
ValueError: 'touch' isn't a valid key for profile (selenium-profiles)
```
?

I did not understand how to solve the problem with following a link from a tab with a changed profile to a new tab opening.

**Comment by ttlns** [comment]
Got what you mean. You're basically not supposed to pass the profile with `driver.profiles.apply(profile["cdp" ])`
The warning ppears when `"options"` is in `profile`. If you don't want the waring, just remove that key with `del profile["options"]`.

Maybe the source-code helps you to understand:
```python
def apply(self, profile:dict):
    """
    apply options after driver allready started
    :param profile: selenium-profiles options
    """
    from selenium_profiles.utils.utils import valid_key
    valid_key(profile.keys(),["cdp", "options"], "profile (selenium-profiles)")

    if "options" in profile.keys():
        warnings.warn('profile["options"] can\'t be applied when driver allready started')

    if "cdp" in profile.keys():
        # noinspection PyUnresolvedReferences
         self.cdp.apply(profile["cdp"])
```

or you could simply do: `driver.profiles.cdp.apply(profile["cdp"])`

I'll might implement in feature automatic apply for profiles. Also, you might need to execute:
```python
from selenium_profiles.scripts.undetected import exec_cdp
exec_cdp(driver)
```
for undetectability on a new tab

**Reply by tnt2k** [reply]
this code gives an error
```
from selenium_profiles.scripts.undetected import exec_cdp
exec_cdp(driver, driver.profiles.cdp_tools)
```


```
    exec_cdp(driver, driver.profiles.cdp_tools)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'utils' object has no attribute 'cdp_tools'
```

**Reply by tnt2k** [reply]
also an observation, if the DOM is modified by means of the js script execution driver by adding, for example, an iframe, the real profile is leaked.

**Reply by ttlns** [reply]
> this code gives an error
> 
> ```
> from selenium_profiles.scripts.undetected import exec_cdp
> exec_cdp(driver, driver.profiles.cdp_tools)
> ```
> 
> ```
>     exec_cdp(driver, driver.profiles.cdp_tools)
>                      ^^^^^^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'utils' object has no attribute 'cdp_tools'
> ```

Nevermind, just leave the second argument out then:)

**Reply by ttlns** [reply]
> don't understand what you mean by that? exec_cdp(driver.profiles.cdp_tools) ?

Oh sorry, just `exec_cdp(driver)` :)

**Reply by tnt2k** [reply]
> > don't understand what you mean by that? exec_cdp(driver.profiles.cdp_tools) ?
> 
> Oh sorry, just `exec_cdp(driver)` :)

the solution you proposed through the operation of Page.addScriptToEvaluateOnNewDocument can work within the same domain, but what about new tabs?

**Reply by ttlns** [reply]
> the solution you proposed through the operation of `Page.addScriptToEvaluateOnNewDocument` can work within the same domain, but what about new tabs?

For now, switch to the tab and then execute it.

**Reply by tnt2k** [reply]
> > the solution you proposed through the operation of `Page.addScriptToEvaluateOnNewDocument` can work within the same domain, but what about new tabs?
> 
> For now, switch to the tab and then execute it.

yes that's understandable, the issue is that opening a new tab on the landing page will lead to a leak.

**Reply by ttlns** [reply]
> yes that's understandable, the issue is that opening a new tab on the landing page will lead to a leak.

exactly.
But tbh not sure how to solve that issue. We'd need some listener at new tab for that
see [crbug#2126](https://bugs.chromium.org/p/chromedriver/issues/detail?id=2126)

**Reply by ttlns** [reply]
@tnt2k Implementation like in https://github.com/puppeteer/puppeteer/issues/3667#issuecomment-562280230 should be possible to fix this issue => working on it :)

Basically attach debugger on every new page

**Comment by ttlns** [comment]
[resource](https://stackoverflow.com/a/33614216/20443541)

> Perhabs you could create an if statement that checks if you are on an mobile device and then loop through all of your links and replace 'target="_blank"' with 'target="_self"'. I think thats fine because you only need this during the developement time on your pc.
> 
> 
>     if(/android|mobile/i.test(navigator.userAgent)){
>        var links = document.getElementsByTagName("a");
>        for(var i = 0; i < links.length; i++){
>           links[i].setAttribute("target","_self");
>        }
>     }

**Comment by ttlns** [comment]
merge into https://github.com/kaliiiiiiiiii/Selenium-Profiles/issues/58
